src/eval.rs

```diff
@@ -204,14 +204,26 @@ pub fn eval_main<'tcx>(tcx: TyCtxt<'tcx>, main_id: DefId, config: MiriConfig) ->
     // Perform the main execution.
     let res: InterpResult<'_, i64> = (|| {
         // Main loop.
-        while ecx.step()? {
+        loop {
+            match ecx.schedule()? {
+                SchedulingAction::ExecuteStep => {
+                    assert!(ecx.step()?, "a terminated thread was scheduled for execution");
+                }
+                SchedulingAction::ExecuteDtors => {
+                    ecx.run_tls_dtors_for_active_thread()?;
```

Comment:
There is one problem with this: `run_tls_dtors_for_active_thread` has its own `step` loop, which means the thread cannot be scheduled away while running TLS dtors, even if it e.g. is blocked on getting a lock from another thread.

I knew that having a separate loop there is a bad idea, but so far I was just too lazy to fix that. ;) Now that's coming back and biting us. We need to switch that to a scheme where we store the information for the next loop iteration in the top frame's `StackPop` info (or so).

But we don't need to solve that problem in this PR. Could you open an issue about this, and add a FIXME here referencing that issue?